File: apps/ios/Orange Cloud/Orange Cloud/ViewModels/ManagedHeadersViewModel.swift
//
//  ManagedHeadersViewModel.swift
//  Orange Cloud
//

import Foundation
import Observation

@Observable
@MainActor
final class ManagedHeadersViewModel {

    private(set) var requestHeaders: [ManagedTransform] = []
    private(set) var responseHeaders: [ManagedTransform] = []
    var isLoading = false
    var loaded = false
    var isMutating = false
    var error: String?
    var didMutate = false

    private let service: ManagedHeaderService
    private let zoneId: String

    init(service: ManagedHeaderService, zoneId: String) {
        self.service = service
        self.zoneId = zoneId
    }

    private func apply(_ transforms: ManagedTransforms) {
        // 接口不保证顺序，按 id 排一下，避免每次刷新行位置跳动
        requestHeaders = (transforms.managedRequestHeaders ?? []).sorted { $0.id < $1.id }
        responseHeaders = (transforms.managedResponseHeaders ?? []).sorted { $0.id < $1.id }
    }

    func load() async {
        isLoading = true
        error = nil
        do {
            apply(try await service.transforms(zoneId: zoneId))
            loaded = true
        } catch is CancellationError {
        } catch let urlError as URLError where urlError.code == .cancelled {
        } catch {
            self.error = error.localizedDescription
        }
        isLoading = false
    }

    func setEnabled(_ item: ManagedTransform, enabled: Bool, isRequest: Bool) async {
        guard !isMutating else { return }
        isMutating = true
        error = nil
        do {
            apply(try await service.setEnabled(
                zoneId: zoneId, id: item.id, enabled: enabled, isRequest: isRequest
            ))
            didMutate.toggle()
        } catch {
            self.error = error.localizedDescription
        }
        isMutating = false
    }
}

File: apps/android/app/src/main/kotlin/jiamin/chen/orangecloud/ui/managedheaders/ManagedHeadersViewModel.kt
package jiamin.chen.orangecloud.ui.managedheaders

import androidx.lifecycle.SavedStateHandle
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import dagger.hilt.android.lifecycle.HiltViewModel
import jiamin.chen.orangecloud.core.auth.AuthRepository
import jiamin.chen.orangecloud.core.auth.Scopes
import jiamin.chen.orangecloud.data.model.ManagedTransform
import jiamin.chen.orangecloud.data.model.ManagedTransforms
import jiamin.chen.orangecloud.data.repository.ManagedHeaderRepository
import kotlinx.coroutines.channels.Channel
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.receiveAsFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import javax.inject.Inject

data class ManagedHeadersUiState(
    val requestHeaders: List<ManagedTransform> = emptyList(),
    val responseHeaders: List<ManagedTransform> = emptyList(),
    val isLoading: Boolean = false,
    val loaded: Boolean = false,
    val isMutating: Boolean = false,
    val missingScope: Boolean = false,
    val canWrite: Boolean = false,
)

sealed interface ManagedHeadersEvent {
    data class Error(val message: String?) : ManagedHeadersEvent
}

@HiltViewModel
class ManagedHeadersViewModel @Inject constructor(
    savedStateHandle: SavedStateHandle,
    private val repository: ManagedHeaderRepository,
    authRepository: AuthRepository,
) : ViewModel() {

    private val zoneId: String = checkNotNull(savedStateHandle["zoneId"])
    private val hasRead = authRepository.hasScope(Scopes.MANAGED_HEADERS_READ)
    private val canWrite = authRepository.hasScope(Scopes.MANAGED_HEADERS_WRITE)

    private val _uiState = MutableStateFlow(
        ManagedHeadersUiState(isLoading = hasRead, missingScope = !hasRead, canWrite = canWrite),
    )
    val uiState: StateFlow<ManagedHeadersUiState> = _uiState.asStateFlow()

    private val eventChannel = Channel<ManagedHeadersEvent>(Channel.BUFFERED)
    val events: Flow<ManagedHeadersEvent> = eventChannel.receiveAsFlow()

    init { if (hasRead) load() }

    /** 接口不保证顺序，按 id 排一下，避免每次刷新行位置跳动 */
    private fun apply(t: ManagedTransforms) = _uiState.update {
        it.copy(
            requestHeaders = t.requestHeaders.orEmpty().sortedBy { h -> h.id },
            responseHeaders = t.responseHeaders.orEmpty().sortedBy { h -> h.id },
            loaded = true,
        )
    }

    fun load() {
        if (!hasRead) return
        viewModelScope.launch {
            _uiState.update { it.copy(isLoading = true) }
            runCatching { repository.transforms(zoneId) }
                .onSuccess { apply(it) }
                .onFailure { eventChannel.send(ManagedHeadersEvent.Error(it.message)) }
            _uiState.update { it.copy(isLoading = false) }
        }
    }

    fun setEnabled(item: ManagedTransform, enabled: Boolean, isRequest: Boolean) {
        if (!canWrite || _uiState.value.isMutating) return
        viewModelScope.launch {
            _uiState.update { it.copy(isMutating = true) }
            runCatching { repository.setEnabled(zoneId, item.id, enabled, isRequest) }
                .onSuccess { apply(it) }
                .onFailure { eventChannel.send(ManagedHeadersEvent.Error(it.message)); load() }
            _uiState.update { it.copy(isMutating = false) }
        }
    }
}
